axoupdater/src/test/helpers.rs

```diff
@@ -0,0 +1,114 @@
+use std::{
+    env::consts::EXE_SUFFIX,
+    path::PathBuf,
+    process::{Command, Stdio},
+};
+
+static RECEIPT_TEMPLATE: &str = r#"{"binaries":[BINARIES],"install_prefix":"INSTALL_PREFIX","provider":{"source":"cargo-dist","version":"0.10.0-prerelease.1"},"source":{"app_name":"APP_NAME","name":"PACKAGE","owner":"axodotdev","release_type":"github"},"version":"VERSION"}"#;
+
+/// Generates an install receipt given the specified fields and returns it as a string.
+pub fn install_receipt(
+    app_name: &str,
+    package: &str,
+    binaries: &[&str],
+    version: &str,
+    prefix: &str,
+) -> String {
+    let binaries = binaries
+        .iter()
+        .map(|name| format!(r#""{name}""#))
+        .collect::<Vec<String>>()
+        .join(", ");
+    RECEIPT_TEMPLATE
+        .replace("BINARIES", &binaries)
+        .replace("PACKAGE", package)
+        .replace("APP_NAME", app_name)
+        .replace("INSTALL_PREFIX", &prefix.replace('\\', "\\\\"))
+        .replace("VERSION", version)
+}
+
+/// Generates an install receipt given the specified fields and writes it to disk at the appropriate location in `config_path`. The path to the new file is returned.
+pub fn write_receipt(
+    app_name: &str,
+    package: &str,
+    binaries: &[&str],
+    version: &str,
+    prefix: &PathBuf,
+    config_path: &PathBuf,
+) -> std::io::Result<PathBuf> {
+    let contents = install_receipt(
+        app_name,
+        package,
+        binaries,
+        version,
+        &prefix.to_string_lossy(),
+    );
+    let receipt_name = config_path.join(format!("{package}-receipt.json"));
+    std::fs::create_dir_all(&config_path)?;
+    std::fs::write(&receipt_name, contents)?;
+
+    Ok(receipt_name)
+}
+
+/// Performs an end-to-end run test for `app_name` within `package`. `bin` is the path to the executable being tested, usually the one from the `CARGO_BIN_EXE_<name>` environment variable.
+/// Rather than returning a value, this function performs assertions.
+/// Because it writes to real files outside a temporary directory, it's highly recommended that this only becalled within CI builds.
+pub fn perform_runtest(
+    app_name: &str,
+    package: &str,
+    bin: &PathBuf,
+    binaries: &[&str],
+    args: &[&str],
+) {
+    let home = homedir::get_my_home().unwrap().unwrap();
+
+    let app_home = &home.join(".cargo").join("bin");
+    let app_path = &app_home.join(format!("{app_name}{EXE_SUFFIX}"));
+
+    #[cfg(target_family = "unix")]
+    let config_path = &home.join(".config").join(app_name);
+    #[cfg(target_family = "windows")]
+    let config_path = &std::env::var("LOCALAPPDATA")
+        .map(std::path::PathBuf::from)
+        .unwrap()
+        .join(app_name);
```

Comment:
Can these be using functions inside the main crate, so they can't diverge?

Reply:
Yeah, I realized these were separate because they originated in cargo-dist, but now they can use existing functions.